File: src/c4/inter/AssignId.java
package inter;
import treewalker.TreeWalker;

/*
 * AssignId ist eine Unterklasse von Assignment und beschreibt 
 * Wertzuweisungen mit einer Variablen auf der linken Seite. 
 * In den Instanzenvariablen id wird die linke Seite, in expr die 
 * rechte Seite der Wertzuweisung abgelegt
 */

public class AssignId extends Assignment {
	Id ident;
	Expr expr;

	public AssignId(Id i, Expr x) {
		ident = i;
		expr = x;
	}
	
	public Id getIdent() {
		return ident;
	}

	public Expr getExpr() {
		return expr;
	}

	public void setExpr(Expr expr) {
		this.expr = expr;
	}

	public <R, P> R walk(TreeWalker<R, P> walker, P arg) {
		return walker.walkAssignIdNode(this, arg);
	}

}

File: src/c4/inter/AssignId.rs
use super::{AssignmentData, AssignmentNode, Expr, Id, Node, NodeData, TreeWalker, Walk};

#[derive(Clone, Debug, PartialEq)]
pub struct AssignId {
    assignment: AssignmentData,
    ident: Id,
    expr: Expr,
}

impl AssignId {
    pub fn new(ident: Id, expr: Expr) -> Self {
        Self {
            assignment: AssignmentData::new(),
            ident,
            expr,
        }
    }

    pub fn get_ident(&self) -> &Id {
        &self.ident
    }
    pub fn get_expr(&self) -> &Expr {
        &self.expr
    }
    pub fn set_expr(&mut self, expr: Expr) {
        self.expr = expr;
    }
}

impl Node for AssignId {
    fn node_data(&self) -> &NodeData {
        &self.assignment.node
    }
}

impl AssignmentNode for AssignId {
    fn assignment_data(&self) -> &AssignmentData {
        &self.assignment
    }
}

impl<P, R> Walk<P, R> for AssignId {
    fn walk<W: TreeWalker<P, R> + ?Sized>(&self, walker: &mut W, arg: P) -> R {
        walker.walk_assign_id_node(self, arg)
    }
}
